lib/Conversion/CalyxNative/CalyxNative.cpp

```diff
@@ -0,0 +1,148 @@
+//===- CalyxNative.cpp - Invoke the native Calyx compiler
+//----------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//===----------------------------------------------------------------------===//
+//
+// Calls out to the native, Rust-based Calyx compiler using the `calyx` binary
+// to run passes.
+//
+//===----------------------------------------------------------------------===//
+
+#include "../PassDetail.h"
+
+#include "circt/Conversion/CalyxNative.h"
+#include "circt/Dialect/Calyx/CalyxEmitter.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Support/FileUtilities.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/Process.h"
+#include "llvm/Support/ToolOutputFile.h"
+
+using namespace mlir;
+using namespace circt;
+
+/// ConversionPatterns.
+
+/// Pass entrypoint.
+
+namespace {
+class CalyxNativePass : public CalyxNativeBase<CalyxNativePass> {
+public:
+  void runOnOperation() override;
+
+private:
+  LogicalResult runOnModule(ModuleOp root);
+};
+} // end anonymous namespace
+
+void CalyxNativePass::runOnOperation() {
+  ModuleOp mod = getOperation();
+  if (failed(runOnModule(mod)))
+    return signalPassFailure();
+}
+
+LogicalResult CalyxNativePass::runOnModule(ModuleOp root) {
+  // User must specify location of the calyx primitive library.
+  if (primitiveLib.empty()) {
+    // XXX(rachitnigam): Probably a bad idea to hard code the name of the
+    // flag. Is there a better way?
+    root.emitError("primitive library not specified. Please specify it using "
```

Comment:
The error message here is really not about the `root` at all and is instead a pass requirement. Is there a better UX way to emit this for the user?

Reply:
Move this into `runOnOperation` and `return signalPassFailure()`?

Reply:
How do you `emitError` without the `root`?

Reply:
I *think* you can also emitError with a Location? Not in front of my laptop. Try looking up InFlightDiagnostic maybe.